crates/core_simd/src/reduction.rs

```diff
@@ -0,0 +1,162 @@
+macro_rules! impl_integer_reductions {
+    { $name:ident, $scalar:ty } => {
+        impl<const LANES: usize> crate::$name<LANES>
+        where
+            Self: crate::LanesAtMost32
+        {
+            /// Horizontal wrapping add.  Returns the sum of the lanes of the vector, with wrapping addition.
+            #[inline]
+            pub fn wrapping_sum(self) -> $scalar {
```

Comment:
should this say `horizontal_` at the front of the name like the others?

Reply:
I left it out of sum and product because I think the names are more self explanatory, since there's no "vertical" sum (and mirror the ones in `std::iter` a bit, too).

Kind of like how it wouldn't really make sense to make it `horizontal_all`.

Reply:
well, these types should implement the Sum and Product traits: https://doc.rust-lang.org/std/iter/trait.Sum.html but if you think it won't be confusing we can skip it for now

Reply:
Unfortunately we can't implement those traits since they only work on iterators.  They get implemented on the iterator `Item`, not the iterator itself.

More specifically you can already do `vector.to_array().sum()` but it won't be SIMD.

Reply:
https://docs.rs/wide/0.6.4/wide/struct.f32x4.html#impl-Sum%3CRHS%3E you just do it like this?

Reply:
That's for accumulating `Iter<Item=f32x4>` into a single `f32x4`, we should probably provide that too.  But there's no way to do horizontal ops with that trait

Reply:
I mean I am still in favor of implementing IntoIterator on vectors, fwiw, precisely for this reason.

Reply:
Right, Sum and Product traits are about "if i have an iterator of you guys, can you put yourselves together?", which is very different from a horizontal sum.

However, the Sum and Product trait *name their methods* `sum` and `product`, so on the assumption that we'll support those traits on these types eventually, we shouldn't name the horizontal sum  and horizontal product methods some name *other* than the ones already used by standard library traits.

So these methods here should be `horizontal_sum` and `horizontal_product`

Reply:
and I understand that there won't be a *literal* namespace clash because one is a method and one is an associated function on a trait but there's a *documentation* and *conceptual* clash for having two things name `sum` or `product` that are doing very different things. It's not normal, it's weird. You wouldn't really name an associated function named `from` and then have it do something *other than* infallibly convert a value from another type into the current type. That would just be a weird thing to do in Rust.

Reply:
There was some resistance the first time I pointed it out as an option but the reason I am noting it is precisely because things like horizontal operations are "you stop caring about the vector as a vector and are reducing it to a single value".

Reply:
@Lokathor I misunderstood your point, I think.  I agree that it's potentially a little confusing about the direction the accumulation occurs, so I renamed them.

@workingjubilee If you're working with iterators over lanes you're forgoing SIMD, so I think it's appropriate to call `to_array()` to make that explicit

Reply:
Mmm, I think there's ways to get around that, but regardless...!

Reply:
To elaborate on how that would look:

An iterator over a 16 byte vector (yon typical SSE/Neon register) that iterated one step (doing whatever) and then had `product()` called on it could have the first lane set to 1 and then issue a horizontal mul instruction. Likewise, sum, but the additive identity is 0. This works because we can extract an ExactSizeIterator from a SIMD vector.

Reply:
Well, first of all that would require different iterators from Sum and Product, which doesn't really make sense.  But I think that's also just not a very good iterator for normal iteration, compared to an array.

Reply:
how about having the iterator be something like:
```rust
struct Iter<SimdType: AsElements> {
    value: SimdType,
    range: Range<usize>,
}

impl<SimdType: AsElements> Iterator for Iter<SimdType> {
    type Item = SimdType::Element;
    fn next(&mut self) -> Option<Self::Item> {
        let index = self.range.next()?;
        unsafe { Some(self.value.get_element_unchecked(index)) }
    }
}

impl<SimdType: AsElements> DoubleEndedIterator for Iter<SimdType> {
    fn next_back(&mut self) -> Option<Self::Item> {
        let index = self.range.next_back()?;
        unsafe { Some(self.value.get_element_unchecked(index)) }
    }
}

trait SumProductHelper<I: Iterator<Item=Self>> {
    fn sum(iter: I) -> Self;
    fn product(iter: I) -> Self;
}

impl<I: Iterator<Item=Self>> SumProductHelper<I> for i32 {
    default fn sum(iter: I) -> Self {
        iter.fold(0i32, Add::add)
    }
    default fn product(iter: I) -> Self {
        iter.fold(1i32, Mul::mul)
    }
}

impl<SimdType: AsElements<Element = Self>> SumProductHelper<Iter<SimdType>> for i32 {
    fn sum(iter: Iter<SimdType>) -> Self {
        iter.value.fill_outside_range(iter.range, 0i32).sum()
    }
    fn product(iter: Iter<SimdType>) -> Self {
        iter.value.fill_outside_range(iter.range, 1i32).product()
    }
}

impl Sum for i32 {
    fn sum<I: Iterator<Item=Self>>(iter: I) -> Self {
        SumProductHelper::sum(iter)
    }
}
```

Reply:
I suppose this would work, but it makes the iterator pretty big and wordy and I'm not convinced this is any better than just doing `as_array` or `as_slice`